src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentStepExecution.java

```diff
@@ -0,0 +1,128 @@
+        }
+        return stepDescriptor;
+    }
+
+    @VisibleForTesting
+    static class Callback extends BodyExecutionCallback {
+
+        private static final Logger LOGGER = Logger.getLogger(Callback.class.getName());
+        private static final long serialVersionUID = 1L;
+
+        private final transient BitbucketDeploymentFactory deploymentFactory;
+        private final transient DeploymentPoster deploymentPoster;
+        private final BitbucketDeploymentEnvironment environment;
+
+        public Callback(BitbucketDeploymentEnvironment environment,
+                        BitbucketDeploymentFactory deploymentFactory,
+                        DeploymentPoster deploymentPoster) {
+            this.environment = environment;
+            this.deploymentFactory = deploymentFactory;
+            this.deploymentPoster = deploymentPoster;
+        }
+
+        @Override
+        public void onStart(StepContext context) {
+            sendNotification(context, DeploymentState.IN_PROGRESS);
+        }
+
+        @Override
+        public void onSuccess(StepContext context, Object result) {
+            sendNotification(context, DeploymentState.SUCCESSFUL);
+            context.onSuccess(result);
+        }
+
+        @Override
+        public void onFailure(StepContext context, Throwable t) {
+            sendNotification(context, DeploymentState.FAILED);
+            context.onFailure(t);
+        }
+
+        private void sendNotification(StepContext context, DeploymentState state) {
+            try {
+                TaskListener listener = context.get(TaskListener.class);
+                if (listener == null) {
+                    LOGGER.info("No task listener in the StepContext. Unable to send deployment information");
+                    return;
+                }
+                Run<?, ?> run = context.get(Run.class);
+                if (run == null) {
+                    LOGGER.info("No run in the StepContext. Unable to send deployment information");
+                    return;
+                }
+                BitbucketRevisionAction revisionAction = run.getAction(BitbucketRevisionAction.class);
+                if (revisionAction == null) {
+                    // Not checked out with a Bitbucket SCM
+                    listener.error("Could not send deployment notification: Run is not using Bitbucket SCM for checkout.");
+                    return;
+                }
+
+                BitbucketDeployment deployment = deploymentFactory.createDeployment(run, environment, state);
+                BitbucketSCMRepository bitbucketSCMRepo = revisionAction.getBitbucketSCMRepo();
+                String revisionSha = revisionAction.getRevisionSha1();
+
+                deploymentPoster.postDeployment(bitbucketSCMRepo, revisionSha, deployment, run, listener);
+            } catch (Exception e) {
+                context.onFailure(e);
```

Comment:
I'm not sure about this. This will fail the job if the deployment notification fails. Maybe we should just log something and proceed. Thoughts?

Reply:
+1. Log with the job, send no status. This is especially true of in-progress notifications- we've had complaints before of a build status posting failing a job, even if the job itself was able to run fine.